File: experiments/problem1_nonperiodicity/analyze_period_two_dominant_shadow_transducer.py
#!/usr/bin/env python3
"""Dominant adjacent-level shadow paths for three-return survivor prefixes.

For phase a, let O_(a,k) be the exact ordinary frontier. For q in O_(a,k),
define the next-level fiber mask

    M_k(q) = {e in {0,1,2,3}: 4q+e in O_(a,k+1)}.

An adjacent pair (q,p), with q at level k and p at level k-1, is dominant when
M_k(q) is a subset of M_(k-1)(p). Then every common base-four digit available
to q is also available to p. Repeating this test along a cylinder gives an
exact finite certificate that a level-k prefix has a same-cylinder shadow at
level k-1.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from collections import Counter, defaultdict
from itertools import product
from typing import Any

PHASES = ("p", "u")
GAPS = (2, 3, 4, 5)
FORBIDDEN = ("uu", "ttttt", "ututtu")
ALLOWED_MASKS = (0b0000, 0b0011, 0b1011, 0b1100, 0b1111)
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
DEFAULT_SCHEDULE_CAP = 64


class DominantShadowLimitError(RuntimeError):
    """Raised before a controlled campaign exceeds its cap."""


def forward_generator(name: str, state: int) -> int:
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(g, state) for g in "tup"}))


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def expected_bits(phase: str, complexity: int) -> int:
    return 2 * complexity if phase == "p" else 2 * complexity - 1


def forced_zero_schedule(state: int, cap: int = DEFAULT_SCHEDULE_CAP) -> str:
    word: list[str] = []
    for _ in range(cap):
        residue = state & 15
        if residue == 7:
            branch = "u"
        elif residue == 11:
            branch = "t"
        else:
            return "".join(word)
        state = forward_generator(
            branch, forward_generator("p", (state - 3) >> 2)
        )
        word.append(branch)
    raise DominantShadowLimitError("forced schedule reached safety cap")


def admissible(word: str) -> bool:
    return not any(factor in word for factor in FORBIDDEN)


def return_extension(gaps: tuple[int, ...], include_final_u: bool) -> str:
    word = "u"
    for index, gap in enumerate(gaps):
        word += "t" * (gap - 1)
        if index < len(gaps) - 1 or include_final_u:
            word += "u"
    return word


def three_return_patterns() -> tuple[tuple[tuple[int, ...], str, str], ...]:
    rows = []
    for gaps in product(GAPS, repeat=3):
        target = return_extension(gaps, False)
        complete = return_extension(gaps, True)
        if admissible(complete):
            rows.append((gaps, target, complete))
    return tuple(rows)


def build_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    levels: list[set[int]] = [set(), {phase_start(phase)}]
    for _ in range(2, maximum_complexity + 1):
        levels.append(
            {child for state in levels[-1] for child in frontier_children(state)}
        )
    return levels


def fiber_mask(levels: list[set[int]], complexity: int, quotient: int) -> int:
    if complexity < 1 or complexity + 1 >= len(levels):
        raise ValueError("fiber level outside built frontier")
    next_level = levels[complexity + 1]
    mask = sum(
        1 << digit for digit in range(4) if 4 * quotient + digit in next_level
    )
    if mask not in ALLOWED_MASKS:
        raise AssertionError("fiber escaped five-mask alphabet")
    return mask


def dominant_path(
    levels: list[set[int]],
    complexity: int,
    current: int,
    shadow: int,
    depth: int,
) -> list[tuple[int, int, int]] | None:
    """Strip a common depth-d cylinder and certify fiber-mask dominance."""
    path: list[tuple[int, int, int]] = []
    for step in range(depth):
        digit = current & 3
        if shadow & 3 != digit:
            return None
        current_quotient = current >> 2
        shadow_quotient = shadow >> 2
        current_level = complexity - 1 - step
        if current_level < 2:
            return None
        current_mask = fiber_mask(levels, current_level, current_quotient)
        shadow_mask = fiber_mask(levels, current_level - 1, shadow_quotient)
        if current_mask & ~shadow_mask:
            return None
        path.append((current_mask, shadow_mask, digit))
        current, shadow = current_quotient, shadow_quotient
    return path


def mask_sequence(
    levels: list[set[int]], complexity: int, state: int, depth: int
) -> tuple[tuple[int, ...], tuple[int, ...]]:
    masks: list[int] = []
    digits: list[int] = []
    for step in range(depth):
        digits.append(state & 3)
        quotient = state >> 2
        level = complexity - 1 - step
        if level < 1:
            raise ValueError("depth exceeds available complexity")
        masks.append(fiber_mask(levels, level, quotient))
        state = quotient
    return tuple(masks), tuple(digits)


def phase_campaign(
    phase: str, maximum_complexity: int, schedule_cap: int
) -> dict[str, Any]:
    levels = build_levels(phase, maximum_complexity)
    patterns = three_return_patterns()
    totals = Counter(
        outputs=sum(len(levels[k]) for k in range(1, maximum_complexity + 1))
    )
    totals["eligible_outputs"] = sum(
        1
        for k in range(1, maximum_complexity + 1)
        for state in levels[k]
        if state & 3 == 3
    )
    pair_counts: Counter[tuple[int, int]] = Counter()
    transition_counts: Counter[tuple[int, int, int]] = Counter()
    path_lengths: Counter[int] = Counter()
    examples: list[dict[str, Any]] = []

    for complexity in range(2, maximum_complexity + 1):
        occurrences: list[tuple[int, int, str, list[tuple[int, ...]]]] = []
        needed_depths: set[int] = set()
        for current in sorted(levels[complexity]):
            if current.bit_length() != expected_bits(phase, complexity):
                raise AssertionError("frontier bit-length law failed")
            if current & 3 != 3:
                continue
            schedule = forced_zero_schedule(current, schedule_cap)
            for cut in range(len(schedule) + 1):
                base = schedule[:cut]
                matches = [
                    gaps
                    for gaps, target, complete in patterns
                    if schedule[cut:].startswith(target)
                    and admissible(base + complete)
                ]
                if matches:
                    occurrences.append((current, cut, base, matches))
                    needed_depths.add(cut + 1)
        if not occurrences:
            continue

        shadow_indexes: dict[int, dict[int, dict[tuple[int, ...], int]]] = {}
        for depth in sorted(needed_depths):
            by_residue: dict[int, dict[tuple[int, ...], int]] = defaultdict(dict)
            modulus = 4**depth
            for shadow in sorted(levels[complexity - 1]):
                if shadow & 3 != 3:
                    continue
                masks, _ = mask_sequence(levels, complexity - 1, shadow, depth)
                residue = shadow % modulus
                by_residue[residue].setdefault(masks, shadow)
            shadow_indexes[depth] = by_residue

        for current, cut, base, matches in occurrences:
            depth = cut + 1
            modulus = 4**depth
            residue = current % modulus
            current_masks, current_digits = mask_sequence(
                levels, complexity, current, depth
            )
            chosen: tuple[int, tuple[int, ...]] | None = None
            for shadow_masks, shadow in shadow_indexes[depth].get(residue, {}).items():
                if all(
                    not (current_mask & ~shadow_mask)
                    for current_mask, shadow_mask in zip(
                        current_masks, shadow_masks
                    )
                ):
                    chosen = (shadow, shadow_masks)
                    break

            for gaps in matches:
                totals["occurrences"] += 1
                if cut > 0:
                    totals["positive_cut_occurrences"] += 1
                totals["maximum_cut"] = max(totals["maximum_cut"], cut)
                if chosen is None:
                    totals["violations"] += 1
                    if len(examples) < 8:
                        examples.append(
                            {
                                "kind": "violation",
                                "complexity": complexity,
                                "state_hex": hex(current),
                                "cut": cut,
                                "base": base,
                                "gaps": list(gaps),
                            }
                        )
                    continue

                shadow, shadow_masks = chosen
                totals["shadowed_occurrences"] += 1
                path_lengths[depth] += 1
                for current_mask, shadow_mask, digit in zip(
                    current_masks, shadow_masks, current_digits
                ):
                    pair_counts[(current_mask, shadow_mask)] += 1
                    transition_counts[(current_mask, shadow_mask, digit)] += 1
                if cut > 0 and len(examples) < 8:
                    examples.append(
                        {
                            "kind": "dominant-shadow",
                            "complexity": complexity,
                            "state_hex": hex(current),
                            "shadow_hex": hex(shadow),
                            "cut": cut,
                            "base": base,
                            "gaps": list(gaps),
                            "mask_path": [
                                [f"0b{a:04b}", f"0b{b:04b}", digit]
                                for a, b, digit in zip(
                                    current_masks, shadow_masks, current_digits
                                )
                            ],
                        }
                    )
        totals["levels_with_occurrences"] += 1

    return {
        "phase": phase,
        "totals": dict(totals),
        "dominant_mask_pairs": {
            f"0b{a:04b}/0b{b:04b}": count
            for (a, b), count in sorted(pair_counts.items())
        },
        "digit_transitions": {
            f"0b{a:04b}/0b{b:04b}/digit-{digit}": count
            for (a, b, digit), count in sorted(transition_counts.items())
        },
        "path_lengths": {
            str(length): count for length, count in sorted(path_lengths.items())
        },
        "examples": examples,
    }


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
    schedule_cap: int = DEFAULT_SCHEDULE_CAP,
) -> dict[str, Any]:
    if not 2 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise DominantShadowLimitError("maximum complexity outside controlled range")
    phases = {
        phase: phase_campaign(phase, maximum_complexity, schedule_cap)
        for phase in PHASES
    }
    combined: Counter[str] = Counter()
    pair_keys: set[str] = set()
    for phase in PHASES:
        combined.update(phases[phase]["totals"])
        pair_keys.update(phases[phase]["dominant_mask_pairs"])
    payload: dict[str, Any] = {
        "status": "exact-dominance-criterion-and-finite-shadow-transducer",
        "maximum_complexity": maximum_complexity,
        "schedule_cap": schedule_cap,
        "admissible_three_return_patterns": len(three_return_patterns()),
        "theorem": {
            "common_digit_simulation": (
                "If adjacent frontier states have current fiber mask contained "
                "in shadow fiber mask, every next base-four digit available to "
                "the current state is also available to the shadow state."
            ),
            "path_certificate": (
                "Repeating the mask-containment test while stripping a shared "
                "cylinder gives an exact certificate that the level-k state has "
                "a same-cylinder level-(k-1) shadow."
            ),
        },
        "phases": phases,
        "combined": dict(combined),
        "observed_pair_alphabet": sorted(pair_keys),
        "scientific_boundary": (
            "The common-digit simulation lemma is exact. Existence of a dominant "
            "shadow for every three-return occurrence is verified only through "
            "the configured complexity; it does not prove the all-depth shadow "
            "inclusion, phase-complexity divergence, exclusion of eventual period "
            "two, or Rule 30 center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity", type=int, default=DEFAULT_MAXIMUM_COMPLEXITY
    )
    parser.add_argument("--schedule-cap", type=int, default=DEFAULT_SCHEDULE_CAP)
    args = parser.parse_args()
    print(json.dumps(
        run_campaign(args.maximum_complexity, args.schedule_cap),
        indent=2,
        sort_keys=True,
    ))


if __name__ == "__main__":
    main()

File: experiments/problem1_nonperiodicity/analyze_period_two_dominant_shadow_transducer.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

namespace {
using u64 = std::uint64_t;
constexpr int MAXK = 25;
constexpr int CAP = 64;
constexpr std::array<int, 4> GAPS{2, 3, 4, 5};

u64 forward(char generator, u64 state) {
    const u64 stepped = state ^ ((state << 1) | (state << 2));
    if (generator == 't') return stepped;
    if (generator == 'u') return stepped ^ 1;
    if (generator == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::runtime_error("unknown generator");
}

std::vector<u64> children(u64 state) {
    std::array<u64, 3> values{
        forward('t', state), forward('u', state), forward('p', state)
    };
    std::sort(values.begin(), values.end());
    return {values.begin(), std::unique(values.begin(), values.end())};
}

int bit_length(u64 value) {
    return value == 0 ? 0 : 64 - __builtin_clzll(value);
}

struct Schedule {
    u64 bits = 0;
    int length = 0;
};

Schedule forced_schedule(u64 state) {
    Schedule result;
    for (int step = 0; step < CAP; ++step) {
        const unsigned residue = state & 15;
        char branch;
        if (residue == 7) branch = 'u';
        else if (residue == 11) branch = 't';
        else return result;
        state = forward(branch, forward('p', (state - 3) >> 2));
        result.bits = (result.bits << 1) | (branch == 'u');
        ++result.length;
    }
    throw std::runtime_error("forced schedule reached cap");
}

struct PrefixKey {
    u64 bits;
    int length;
    bool operator==(const PrefixKey& other) const {
        return bits == other.bits && length == other.length;
    }
};

struct PrefixHash {
    std::size_t operator()(const PrefixKey& key) const noexcept {
        u64 value = key.bits ^ (u64(key.length) << 57);
        value ^= value >> 30;
        value *= 0xbf58476d1ce4e5b9ULL;
        value ^= value >> 27;
        value *= 0x94d049bb133111ebULL;
        value ^= value >> 31;
        return static_cast<std::size_t>(value);
    }
};

PrefixKey prefix(const Schedule& schedule, int length) {
    return {schedule.bits >> (schedule.length - length), length};
}

std::string decode(PrefixKey key) {
    std::string word;
    for (int bit = key.length - 1; bit >= 0; --bit) {
        word.push_back(((key.bits >> bit) & 1) ? 'u' : 't');
    }
    return word;
}

bool admissible(const std::string& word) {
    return word.find("uu") == std::string::npos
        && word.find("ttttt") == std::string::npos
        && word.find("ututtu") == std::string::npos;
}

std::string extension(const std::array<int, 3>& gaps, bool include_final_u) {
    std::string word = "u";
    for (int index = 0; index < 3; ++index) {
        word += std::string(gaps[index] - 1, 't');
        if (index < 2 || include_final_u) word += 'u';
    }
    return word;
}

struct Pattern {
    std::array<int, 3> gaps;
    std::string target;
    std::string complete;
};

std::vector<Pattern> patterns() {
    std::vector<Pattern> output;
    for (int first : GAPS) {
        for (int second : GAPS) {
            for (int third : GAPS) {
                std::array<int, 3> gaps{first, second, third};
                const std::string target = extension(gaps, false);
                const std::string complete = extension(gaps, true);
                if (admissible(complete)) {
                    output.push_back({gaps, target, complete});
                }
            }
        }
    }
    return output;
}

bool schedule_has_at(const Schedule& schedule, int cut, const std::string& target) {
    if (cut + static_cast<int>(target.size()) > schedule.length) return false;
    for (int index = 0; index < static_cast<int>(target.size()); ++index) {
        const int shift = schedule.length - 1 - (cut + index);
        const char observed = ((schedule.bits >> shift) & 1) ? 'u' : 't';
        if (observed != target[index]) return false;
    }
    return true;
}

bool contains(const std::vector<u64>& values, u64 target) {
    return std::binary_search(values.begin(), values.end(), target);
}

unsigned fiber(const std::vector<u64>& next_level, u64 quotient) {
    unsigned mask = 0;
    for (int digit = 0; digit < 4; ++digit) {
        if (contains(next_level, 4 * quotient + digit)) mask |= 1U << digit;
    }
    return mask;
}

struct Totals {
    u64 outputs = 0;
    u64 eligible = 0;
    u64 occurrences = 0;
    u64 positive = 0;
    u64 shadowed = 0;
    u64 violations = 0;
    u64 states = 0;
    int maximum_cut = 0;
    std::map<std::pair<unsigned, unsigned>, u64> pairs;
    std::map<std::tuple<unsigned, unsigned, int>, u64> transitions;
};

bool dominant(
    const std::vector<std::vector<u64>>& levels,
    int complexity,
    u64 current,
    u64 shadow,
    int depth,
    std::vector<std::tuple<unsigned, unsigned, int>>& path
) {
    path.clear();
    for (int step = 0; step < depth; ++step) {
        const int digit = current & 3;
        if (static_cast<int>(shadow & 3) != digit) return false;
        const u64 current_quotient = current >> 2;
        const u64 shadow_quotient = shadow >> 2;
        const int current_level = complexity - 1 - step;
        if (current_level < 2) return false;
        const unsigned current_mask = fiber(
            levels[current_level + 1], current_quotient
        );
        const unsigned shadow_mask = fiber(
            levels[current_level], shadow_quotient
        );
        path.emplace_back(current_mask, shadow_mask, digit);
        if (current_mask & ~shadow_mask) return false;
        current = current_quotient;
        shadow = shadow_quotient;
    }
    return true;
}

Totals run_phase(
    char phase,
    const std::vector<std::vector<u64>>& levels,
    const std::vector<Pattern>& pattern_rows
) {
    Totals totals;
    totals.outputs = 1;
    totals.eligible = phase == 'p' ? 1 : 0;

    for (int complexity = 2; complexity <= MAXK; ++complexity) {
        const auto& current_level = levels[complexity];
        const auto& previous_level = levels[complexity - 1];
        std::unordered_map<PrefixKey, std::vector<u64>, PrefixHash> prefix_map;
        prefix_map.reserve(previous_level.size());
        for (u64 shadow : previous_level) {
            if ((shadow & 3) != 3) continue;
            const Schedule schedule = forced_schedule(shadow);
            for (int length = 0; length <= schedule.length; ++length) {
                prefix_map[prefix(schedule, length)].push_back(shadow);
            }
        }

        u64 level_occurrences = 0;
        u64 level_violations = 0;
        for (u64 current : current_level) {
            const int expected = phase == 'p' ? 2 * complexity : 2 * complexity - 1;
            if (bit_length(current) != expected) {
                throw std::runtime_error("frontier bit-length law failed");
            }
            if ((current & 3) != 3) continue;
            ++totals.eligible;
            const Schedule schedule = forced_schedule(current);
            bool state_has_occurrence = false;

            for (int cut = 0; cut <= schedule.length; ++cut) {
                const PrefixKey key = prefix(schedule, cut);
                const std::string base = decode(key);
                int matches = 0;
                for (const Pattern& pattern : pattern_rows) {
                    if (schedule_has_at(schedule, cut, pattern.target)
                        && admissible(base + pattern.complete)) {
                        ++matches;
                    }
                }
                if (!matches) continue;
                state_has_occurrence = true;

                bool found = false;
                std::vector<std::tuple<unsigned, unsigned, int>> chosen_path;
                const auto found_prefix = prefix_map.find(key);
                if (found_prefix != prefix_map.end()) {
                    for (u64 shadow : found_prefix->second) {
                        std::vector<std::tuple<unsigned, unsigned, int>> path;
                        if (dominant(
                                levels, complexity, current, shadow, cut + 1, path
                            )) {
                            found = true;
                            chosen_path = std::move(path);
                            break;
                        }
                    }
                }

                for (int match = 0; match < matches; ++match) {
                    ++totals.occurrences;
                    ++level_occurrences;
                    if (cut > 0) ++totals.positive;
                    totals.maximum_cut = std::max(totals.maximum_cut, cut);
                    if (found) {
                        ++totals.shadowed;
                        for (const auto& [current_mask, shadow_mask, digit] : chosen_path) {
                            ++totals.pairs[{current_mask, shadow_mask}];
                            ++totals.transitions[{current_mask, shadow_mask, digit}];
                        }
                    } else {
                        ++totals.violations;
                        ++level_violations;
                    }
                }
            }
            if (state_has_occurrence) ++totals.states;
        }
        totals.outputs += current_level.size();
        if (level_occurrences || level_violations) {
            std::cout << "phase=" << phase
                      << " k=" << complexity
                      << " occ=" << level_occurrences
                      << " viol=" << level_violations
                      << " prefix_keys=" << prefix_map.size() << '\n';
        }
    }
    return totals;
}
}  // namespace

int main() {
    const std::vector<Pattern> pattern_rows = patterns();
    std::cout << "patterns=" << pattern_rows.size() << '\n';
    Totals total;

    for (char phase : {'p', 'u'}) {
        std::vector<std::vector<u64>> levels(MAXK + 1);
        levels[1] = {phase == 'p' ? 3ULL : 1ULL};
        for (int complexity = 2; complexity <= MAXK; ++complexity) {
            std::vector<u64> next;
            next.reserve(levels[complexity - 1].size() * 3);
            for (u64 state : levels[complexity - 1]) {
                const std::vector<u64> values = children(state);
                next.insert(next.end(), values.begin(), values.end());
            }
            std::sort(next.begin(), next.end());
            next.erase(std::unique(next.begin(), next.end()), next.end());
            levels[complexity] = std::move(next);
        }

        const Totals phase_totals = run_phase(phase, levels, pattern_rows);
        std::cout << "summary phase=" << phase
                  << " outputs=" << phase_totals.outputs
                  << " eligible=" << phase_totals.eligible
                  << " occ=" << phase_totals.occurrences
                  << " positive=" << phase_totals.positive
                  << " shadowed=" << phase_totals.shadowed
                  << " viol=" << phase_totals.violations
                  << " states=" << phase_totals.states
                  << " maxcut=" << phase_totals.maximum_cut << '\n';
        for (const auto& [pair, count] : phase_totals.pairs) {
            std::cout << "pair phase=" << phase
                      << " current=" << pair.first
                      << " shadow=" << pair.second
                      << " count=" << count << '\n';
        }

        total.outputs += phase_totals.outputs;
        total.eligible += phase_totals.eligible;
        total.occurrences += phase_totals.occurrences;
        total.positive += phase_totals.positive;
        total.shadowed += phase_totals.shadowed;
        total.violations += phase_totals.violations;
        total.states += phase_totals.states;
        total.maximum_cut = std::max(total.maximum_cut, phase_totals.maximum_cut);
        for (const auto& [pair, count] : phase_totals.pairs) {
            total.pairs[pair] += count;
        }
        for (const auto& [transition, count] : phase_totals.transitions) {
            total.transitions[transition] += count;
        }
    }

    std::cout << "total outputs=" << total.outputs
              << " eligible=" << total.eligible
              << " occ=" << total.occurrences
              << " positive=" << total.positive
              << " shadowed=" << total.shadowed
              << " viol=" << total.violations
              << " states=" << total.states
              << " maxcut=" << total.maximum_cut << '\n';
    for (const auto& [pair, count] : total.pairs) {
        std::cout << "pair current=" << pair.first
                  << " shadow=" << pair.second
                  << " count=" << count << '\n';
    }
    for (const auto& [transition, count] : total.transitions) {
        std::cout << "triple current=" << std::get<0>(transition)
                  << " shadow=" << std::get<1>(transition)
                  << " digit=" << std::get<2>(transition)
                  << " count=" << count << '\n';
    }
    return total.violations == 0 ? 0 : 2;
}
